pyquil/reference_simulator.py

```diff
@@ -0,0 +1,234 @@
+        self.wf[0] = complex(1.0, 0)
+
+    def sample_bitstrings(self, n_samples):
+        """
+        Sample bitstrings from the distribution defined by the wavefunction.
+
+        Qubit 0 is at ``out[:, 0]``.
+
+        :param n_samples: The number of bitstrings to sample
+        :return: An array of shape (n_samples, n_qubits)
+        """
+        if self.rs is None:
+            raise ValueError("You have tried to perform a stochastic operation without setting the "
+                             "random state of the simulator. Might I suggest using a PyQVM object?")
+        probabilities = np.abs(self.wf) ** 2
+        possible_bitstrings = all_bitstrings(self.n_qubits)
+        inds = self.rs.choice(2 ** self.n_qubits, n_samples, p=probabilities)
+        bitstrings = possible_bitstrings[inds, :]
+        bitstrings = np.flip(bitstrings, axis=1)  # qubit ordering: 0 on the left.
+        return bitstrings
+
+    def do_gate(self, gate: Gate):
+        """
+        Perform a gate.
+
+        :return: ``self`` to support method chaining.
+        """
+        unitary = lifted_gate(gate=gate, n_qubits=self.n_qubits)
+        self.wf = unitary.dot(self.wf)
+        return self
+
+    def do_measurement(self, qubit: int) -> int:
+        """
+        Measure a qubit, collapse the wavefunction, and return the measurement result.
+
+        :param qubit: Index of the qubit to measure.
+        :return: measured bit
+        """
+        if self.rs is None:
+            raise ValueError("You have tried to perform a stochastic operation without setting the "
+                             "random state of the simulator. Might I suggest using a PyQVM object?")
+        # lift projective measure operator to Hilbert space
+        # prob(0) = <psi P0 | P0 psi> = psi* . P0* . P0 . psi
+        measure_0 = lifted_gate_matrix(matrix=P0, qubit_inds=[qubit], n_qubits=self.n_qubits)
+        proj_psi = measure_0 @ self.wf
+        prob_zero = np.conj(proj_psi).T @ proj_psi
+
+        # generate random number to 'roll' for measure
+        if self.rs.uniform() < prob_zero:
+            # decohere state using the measure_0 operator
+            unitary = measure_0 @ (np.eye(2 ** self.n_qubits) / np.sqrt(prob_zero))
+            self.wf = unitary.dot(self.wf)
+            return 0
+        else:  # measure one
+            measure_1 = lifted_gate_matrix(matrix=P1, qubit_inds=[qubit], n_qubits=self.n_qubits)
+            unitary = measure_1 @ (np.eye(2 ** self.n_qubits) / np.sqrt(1 - prob_zero))
+            self.wf = unitary.dot(self.wf)
+            return 1
+
+    def expectation(self, operator: Union[PauliTerm, PauliSum]):
+        """
+        Compute the expectation of an operator.
+
+        :param operator: The operator
+        :return: The operator's expectation value
+        """
+        if not isinstance(operator, PauliSum):
+            operator = PauliSum([operator])
+
+        sum = 0
```

Comment:
`return sum([_term_expectation(self.wf, term, n_qubits=self.n_qubits) for termin operator])`

Reply:
no list